### PR TITLE
[OpenLSTO] Add Builder

### LINKED ISSUE #3376: OpenLSTO
Would be great to have a JLL package for https://github.com/M2DOLab/OpenLSTO. I tried many things but couldn't get BB to work on an M1, Linux Arm VM on the M1, JuliaHub and AWS. I am yet to try a basic Linux machine but unfortunately I don't have to access to one at the moment, so here I am...

### PR DESCRIPTION
**Pull request by Wimmerer**
Will close #3376 once completed. I think I've got one of the 3 "libraries" built, but I'd like to get the makefile checked before I proceed. I don't have a great way to test that what I built will work.

Questions: 

~~Should the main product be `m2do_fea.so` or `m2do_fea`?~~ Resolved: use `$(dlext)`.

One makefile per "library" and then one more in the `OpenLSTO` directory that just `include`s the others?

### CODE REVIEW

**Comment by Wimmerer** [comment]
I think most of those (except M1) are failing because I hardcoded `.so` as the extension. Should I leave the extension off?

**Comment by Wimmerer** [comment]
@giordano On a scale from cmake-10 how ok is this makefile? 

Also I disabled Apple x86, it was missing `omp.h`. Eigen doesn't exist for M1, but it'd be nice to reenable x86.

**Comment by giordano** [comment]
> @giordano On a scale from cmake-10 how ok is this makefile?

8 :slightly_smiling_face: 

> Also I disabled Apple x86, it was missing omp.h. Eigen doesn't exist for M1, but it'd be nice to reenable x86.

Add `CompilerSupportLibraries_jll`.  Also, `Eigen` should really be rebuilt for `AnyPlatform` (maybe v3.3.9), the current builder probably predates `AnyPlatform`.  I see we have some packages that [depend on `Eigen_jll`](https://juliahub.com/ui/Packages/Eigen_jll/tBNLX/3.3.7+1?t=2) :persevere:

File: O/OpenLSTO/bundled/makefile_FEA [hunk 1]
```diff
@@ -0,0 +1,49 @@
+
+CPPFLAGS = -std=c++14 -Wall -fopenmp -w -O3 -fPIC -fopenmp
```
Comment (giordano) [inline O/OpenLSTO/bundled/makefile_FEA:2]: Note that CPPFLAGS are typically flags for the preprocessor. Here you likely want to use CXXFLAGS

File: O/OpenLSTO/bundled/patches/include.patch [hunk 2]
```diff
@@ -0,0 +1,366 @@
+diff --git a/M2DO_LSM/include/boundary.h b/M2DO_LSM/include/boundary.h
```
Comment (giordano) [inline O/OpenLSTO/bundled/patches/include.patch:1]: Wh... why do we need all of this? 😕

File: O/OpenLSTO/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+platforms = expand_cxxstring_abis(supported_platforms(; experimental=true))
+filter!(p -> !(Sys.isapple(p) && arch(p) == "aarch64"), platforms)
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("m2do_fea", :m2do_fea)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency("Eigen_jll"),
```
Comment (giordano) [inline O/OpenLSTO/build_tarballs.jl:42]: BTW, this is a BuildDependency

File: O/OpenLSTO/bundled/makefile_FEA (hunk 1 above)
Comment (Wimmerer) [reply O/OpenLSTO/bundled/makefile_FEA:2]: Oops, thought I had fixed that.

File: O/OpenLSTO/bundled/patches/include.patch (hunk 2 above)
Comment (Wimmerer) [reply O/OpenLSTO/bundled/patches/include.patch:1]: You mean the whole patch? I have no idea... I got it to compile. Without the patch I get tons of missing definitions etc. I sort of adapted what someone did to build this from Python.

File: O/OpenLSTO/bundled/makefile_FEA (hunk 1 above)
Comment (giordano) [reply O/OpenLSTO/bundled/makefile_FEA]: Also, `-Wall` and `-w` are kind of opposite?

File: O/OpenLSTO/bundled/patches/include.patch (hunk 2 above)
Comment (Wimmerer) [reply O/OpenLSTO/bundled/patches/include.patch:1]: If there's an easier fix I'd love to do it.

Comment (giordano) [reply O/OpenLSTO/bundled/patches/include.patch:1]: How can this work at all if all these includes are necessary? 😕

Comment (Wimmerer) [reply O/OpenLSTO/bundled/patches/include.patch:1]: They might not be. Previously they had the `.cpp` files included into the `.h` files, but it was "Undefined" sort of errors all over the place. I'm certain there's a more succinct way, I just went along including whatever the compiler complained about.

Comment (Wimmerer) [reply O/OpenLSTO/bundled/patches/include.patch:1]: So this is definitely the "noob fix".

**Comment by Wimmerer** [comment]
I've still got 1 more subpackage to build, I'll finish that up later today.

**Comment by Wimmerer** [comment]
It's picking up Eigen now, but not finding `omp.h`. `/workspace/destdir/include/eigen3/Eigen/Core:266:10: fatal error: 'omp.h' file not found`

**Comment by giordano** [comment]
> On a scale from cmake-10 how ok is this makefile?

To get a better score, `all` should be the first rule, so that'd be the default of `make` without arguments

File: O/OpenLSTO/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+install_license ${WORKSPACE}/srcdir/OpenLSTO/LICENSE
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = expand_cxxstring_abis(supported_platforms(; experimental=true); skip=p->false)
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("m2do_fea", :m2do_fea)
```
Comment (mohamed82008) [inline O/OpenLSTO/build_tarballs.jl:44]: what about the other libraries?

Comment (Wimmerer) [reply O/OpenLSTO/build_tarballs.jl]: Oops fixed :)

**Comment by Wimmerer** [comment]
Oops I need to modify the makefiles. On mobile, please don't merge.

**Comment by Wimmerer** [comment]
Alright this is ready to go.

**Comment by mohamed82008** [comment]
Thank you @Wimmerer! This was wat more difficult than I thought it would be, definitely couldn't have done it on my own 😅

**Pull request merged**